### PR TITLE
KAFKA-4830: Augment KStream.print() to allow users pass in extra parameters in the printed string

### PR DESCRIPTION
**Pull request by jedichien**
I extend `KStream#print()` to `KStream#print(KeyValueMapper<K, V, String>)`.
So I add the following methods :
1. `KStream#print(KeyValueMapper<K, V, String>)`
2. `KStream#print(KeyValueMapper<K, V, String>, String streamName)`
3. `KStream#print(KeyValueMapper<K, V, String>, Serde<K>, Serde<V>)`
4. `KStream#print(KeyValueMapper<K, V, String>, Serde<K>, Serde<V>, String streamName)`

### CODE REVIEW

**Comment by bbejeck** [comment]
@jedichien - there are few files in conflict, can you rebase?

**Comment by bbejeck** [comment]
@jedichien - the build is failing for checkstyle errors `KStreamImpl.java:185:20: ',' is not followed by whitespace.`  Also, still 3 files with conflicts.

**Comment by jedichien** [comment]
@bbejeck Thank for reminding me :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 1]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
                final String streamName);
 
     /**
+     * Print the customized output with {@code System.out}.
+     * <p>
+     * The default serde will be use to deserialize key or value if type is {@code byte[]}.
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: EDIT: comment good as is

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 2]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
...
+     * The user provided {@link KeyValueMapper} which customizes output is used to print with {@code System.out}
+     * <p>
+     * The example below shows the way to customize output data.
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:334]: Might want to mention any custom Objects will need to have a good `toString` implementation to get meaningful information.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 3]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
...
+     */
+    void print(final KeyValueMapper<? super K, ? super V, String> mapper);
+
+    /**
+     * Print the customized output with {@code System.out}.
+     * <p>
+     * The default serde will be used to deserialize key or value if type is {@code byte[]}.
+     * The user provided {@link KeyValueMapper} which customizes output is used to print with {@code System.out}
+     * <p>
+     * The example below shows the way to customize output data.
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:356]: EDIT: just mention overriding `toString` for good info.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 4]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
...
+     * @param streamName The given name which labels output will be printed.
+     */
+    void print(final KeyValueMapper<? super K, ? super V, String> mapper, final String streamName);
+
+    /**
+     * Print the customized output with {@code System.out}.
+     * <p>
+     * The user provided {@link KeyValueMapper} which customizes output is used to print with {@code System.out}
+     * The provided serde will be use to deserialize key or value if type is {@code byte[]}.
+     * <p>
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:378]: DItto from above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 5]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
...
+     *     }
+     * };
+     * }</pre>
+     * <p>
+     * The provided KeyValueMapper's mapped value type must be {@code String}.
+     *
+     * @param mapper a {@link KeyValueMapper} that computes output type {@code String}.
+     * @param keySerde a {@link Serde<K>} used to deserialize key if type is {@code byte[]}.
+     * @param valSerde a {@link Serde<V>} used to deserialize value if type is {@code byte[]}.
+     */
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:397]: Ditto from above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 6]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
...
+    void print(final KeyValueMapper<? super K, ? super V, String> mapper, final Serde<K> keySerde, final Serde<V> valSerde);
+
+    /**
+     * Print the customized output with {@code System.out}.
+     * <p>
+     * The user provided {@link KeyValueMapper} which customizes output is used to print with {@code System.out}.
+     * The provided serde will be use to deserialize key or value if type is {@code byte[]}.
+     * <p>
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:405]: DITTO from above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java [hunk 7]
```diff
@@ -22,7 +22,7 @@
 
     private final String streamName;
     private final PrintWriter printWriter;
-
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: nit: is this space intentional?

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java [hunk 8]
```diff
@@ -34,13 +34,30 @@
...
+    public PrintForeachAction(final PrintWriter printWriter, KeyValueMapper<? super K, ? super V, String> mapper, final String streamName) {
         this.printWriter = printWriter;
+        this.mapper = mapper;
         this.streamName = streamName;
     }
 
     @Override
     public void apply(final K key, final V value) {
-        final String data = String.format("[%s]: %s, %s", streamName, key, value);
+        final String data = (mapper == null) ? String.format("[%s]: %s, %s", streamName, key, value) : String.format("[%s]: %s", streamName, mapper.apply(key, value));
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: Just a thought here.  If the `PrintForeachAction(final PrintWriter printWriter, final String streamName)` constructor is used instead of passing null, we could pass in an identity `KeyValueMapper` instance.  This would simplify the code on line 59 by removing the null check.  

WDYT?

**Comment by bbejeck** [comment]
@jedichien thanks for cleaning up the PR.  I've left some minor comments.

Has KIP-160 gone through the community discussion?  Also it looks like some details could be filled out in KIP-160.

\cc @mjsax @guozhangwang @dguy @enothereska  for second review

**Comment by guozhangwang** [comment]
Sometimes if the KIP itself is very trivial people can choose to start the voting thread together with the discussion thread.

For example:
http://mail-archives.apache.org/mod_mbox/kafka-dev/201702.mbox/%3CCAE1jLMO3NzM%2BAj1vf-io9-cnGZwR0y%3DaauAPwQEBP2JAR28Gsg%40mail.gmail.com%3E

https://www.mail-archive.com/dev@kafka.apache.org/msg72060.html

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 2 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:334]: @bbejeck I think no need to mention this due to only accept String type in KeyValueMapper's output part.

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java (hunk 8 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: @bbejeck I think it will be more complex because I have to implement KeyValueMapper in `PrintForeach(final PrintWriter printWriter, final String streamName)`.

**Comment by jedichien** [comment]
KIP-160 is here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-160%3A+Augment+KStream.print%28%29+to+allow+users+pass+in+extra+parameters+in+the+printed+string

**Comment by bbejeck** [comment]
Thanks for the KIP!

+1

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5652/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5638/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by mjsax** [comment]
What's the status of this. Can this be merged \cc @guozhangwang @dguy

**Comment by jedichien** [comment]
Hi, can someone review my code ?
This request blocks me so long.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 9]
```diff
@@ -326,6 +326,100 @@ void print(final Serde<K> keySerde,
...
+     *     }
+     * };
+     * }</pre>
+     * <p>
+     * The KeyValueMapper's mapped value type must be {@code String}.
+     *
+     * @param mapper a {@link KeyValueMapper} that computes output type {@code String}.
+     * @param streamName The given name which labels output will be printed.
+     */
+    void print(final KeyValueMapper<? super K, ? super V, String> mapper, final String streamName);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: Maybe we should rename `streamName` to `label` or `printPrefix`, or something else. i.e., it isn't necessarily the streamName, it is more of print label, tag, prefix etc. Same in other overloads

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java (hunk 8 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: +1 on not passing null.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 10]
```diff
@@ -177,63 +177,96 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
 
     @Override
     public void print() {
-        print(null, null, null);
+        print(null, null, null, null);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: I'd prefer that we pass in `this.name` here instead of `null`.  In the method it delegates to we should do `Objects.requireNonNull(streamName, "streamName can't be null");`

Same for other methods that are passing null

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 11]
```diff
@@ -177,63 +177,96 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
...
+    public void print(KeyValueMapper<? super K, ? super V, String> mapper, Serde<K> keySerde, Serde<V> valSerde) {
+        print(mapper, keySerde, valSerde, null);
+    }
+
+    @Override
+    public void print(KeyValueMapper<? super K, ? super V, String> mapper, Serde<K> keySerde, Serde<V> valSerde, String streamName) {
         String name = topology.newName(PRINTING_NAME);
         streamName = (streamName == null) ? this.name : streamName;
-        topology.addProcessor(name, new KStreamPrint<>(new PrintForeachAction(null, streamName), keySerde, valSerde), this.name);
+        topology.addProcessor(name, new KStreamPrint<>(new PrintForeachAction(null, mapper, streamName), keySerde, valSerde), this.name);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Not done as part of the PR, but... Can we pass `new PrintWriter(System.out)` here instead of `null`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 12]
```diff
@@ -177,63 +177,96 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
...
     @Override
     public void writeAsText(String filePath) {
-        writeAsText(filePath, null, null, null);
+        writeAsText(filePath, null, null, null, null);
     }
 
     @Override
     public void writeAsText(String filePath, String streamName) {
-        writeAsText(filePath, streamName, null, null);
+        writeAsText(filePath, streamName, null, null, null);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: I'd really like to discourage passing `null`. We can have a `KeyValueMapper` instance that we pass here and also throw an exception in the method that is delegated to if the `KeyValueMapper` is `null`.

Same elsewhere

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 13]
```diff
@@ -177,63 +177,96 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
...
+    }
+
+    @Override
+    public void writeAsText(final String filePath, final Serde<K> keySerde, final Serde<V> valSerde, final KeyValueMapper<? super K, ? super V, String> mapper) {
+        writeAsText(filePath, null, keySerde, valSerde, mapper);
+    }
+
+    @Override
+    public void writeAsText(final String filePath, String streamName, final Serde<K> keySerde, final Serde<V> valSerde, final KeyValueMapper<? super K, ? super V, String> mapper) {
         Objects.requireNonNull(filePath, "filePath can't be null");
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: As i said above, we should `requireNonNull(mapper, ...)`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 14]
```diff
@@ -177,63 +177,96 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
...
             throw new TopologyBuilderException("filePath can't be an empty string");
         }
         String name = topology.newName(PRINTING_NAME);
         streamName = (streamName == null) ? this.name : streamName;
         try {
-            PrintWriter printWriter = null;
-            printWriter = new PrintWriter(filePath, StandardCharsets.UTF_8.name());
-            topology.addProcessor(name, new KStreamPrint<>(new PrintForeachAction(printWriter, streamName), keySerde, valSerde), this.name);
+            PrintWriter printWriter = new PrintWriter(filePath, StandardCharsets.UTF_8.name());
+            topology.addProcessor(name, new KStreamPrint<>(new PrintForeachAction(printWriter, mapper, streamName), keySerde, valSerde), this.name);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: `PrintForEachAction<>` to remove warning. Also in the `print` method

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java [hunk 15]
```diff
@@ -34,13 +34,30 @@
      * @param streamName The given name will be printed.
      */
     public PrintForeachAction(final PrintWriter printWriter, final String streamName) {
+        this(printWriter, null, streamName);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: replace `null` with a `KeyValueMapper`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java [hunk 16]
```diff
@@ -34,13 +34,30 @@
...
+     * then it would close {@code System.out} output stream.
+     * <p>
+     * Afterall, not to pass in {@code PrintWriter(System.out)} but {@code null} instead.
+     *
+     * @param printWriter Use {@code System.out.println} if {@code null}.
+     * @param mapper The mapper which can allow user to customize output will be printed.
+     * @param streamName The given name will be printed.
+     */
+    public PrintForeachAction(final PrintWriter printWriter, KeyValueMapper<? super K, ? super V, String> mapper, final String streamName) {
         this.printWriter = printWriter;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: Mostly not part of this PR, but.. If we validate all params are non-null here then we can remove the null checks elsewhere.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java [hunk 17]
```diff
@@ -82,9 +83,42 @@ public void testPrintKeyValueWithName() {
...
+
+    @Test
+    public void testPrintKeyValueMapperWithName() {
+        final KeyValueMapper<Integer, String, String> mapper = new KeyValueMapper<Integer, String, String>() {
+            @Override
+            public String apply(Integer key, String value) {
+                return String.format("(%d, %s)", key, value);
+            }
+        };
+        final KStreamPrint<Integer, String> kStreamPrint = new KStreamPrint<>(new PrintForeachAction(printWriter, mapper, "test-stream"), intSerd, stringSerd);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: nit: `PrintForEachAction<>`

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java [hunk 18]
```diff
@@ -82,9 +83,42 @@ public void testPrintKeyValueWithName() {
...
+        final KStreamBuilder builder = new KStreamBuilder();
+        final KStream<Integer, String> stream = builder.stream(intSerd, stringSerd, topicName);
+        stream.process(kStreamPrint);
+
+        driver = new KStreamTestDriver(builder);
+        for (KeyValue<Integer, String> record: inputRecords) {
+            driver.process(topicName, record.key, record.value);
+        }
+        printWriter.flush();
+        final String[] flushOutDatas = new String(byteOutStream.toByteArray(), Charset.forName("UTF-8")).split("\\r*\\n");
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: `flushOutDatas` -> `results`

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java [hunk 19]
```diff
@@ -82,9 +83,42 @@ public void testPrintKeyValueWithName() {
...
-        final String[] flushOutDatas = new String(byteOutStream.toByteArray(), Charset.forName("UTF-8")).split("\n");
+        final String[] flushOutDatas = new String(byteOutStream.toByteArray(), Charset.forName("UTF-8")).split("\\r*\\n");
         for (int i = 0; i < flushOutDatas.length; i++) {
-            assertEquals(flushOutDatas[i], expectedResult[i]);
+            assertEquals(expectedResult[i], flushOutDatas[i]);
+        }
+    }
+
+    @Test
+    public void testPrintKeyValueMapperWithName() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: i find it useful for the test name to describe what it is doing. something like:
`shouldPrintStreamWithProvidedKeyValueMapper`

**Review by dguy — CHANGES_REQUESTED** [review]
Thanks @jedichien for the patch. I've left a few comments. One general thing, we should avoid passing `null` around as much as possible. I know that this exists to a certain degree already, but i'd like to see this pattern change such that we rarely pass `null`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 9 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: I think, label, is good :)

Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: but I notice that `streamName` is origin, so not in my PR but also in others like `KTableImpl.print`.
So, I will change others too if I rename it.
WDYT @dguy

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java (hunk 15 above)
Comment (jedichien) [inline streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: Only `KTableImpl.print` and `KTableImpl.writeAsText` use this `PrintForeachAction(final PrintWriter printWriter, final String streamName)`, so I decide to delete this one.
It will make code more clean :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java (hunk 16 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: I already have null check for KeyValueMapper in `PrintForeachAction.apply` and will use default format if null.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 11 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: No, we should not pass `System.out`, because it will lead to close output stream of console.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 12 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: The reason is if I pass in KeyValueMapper instead of passing null, I should create a instance in `print()` and `writeAsText()` which is 
```
final KeyValueMapper<K, V, String> mapper = new KeyValueMapper<K, V, String>() {
          @Override
          public String apply(K key, V value) {
              return String.format("%s, %s", key, value);
          }
 };
```
and I think it will make code unclean.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 10 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Yes, I replace it with passing `this.name` for label :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 14 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: done :)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java (hunk 17 above)
Comment (jedichien) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: done :)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java (hunk 18 above)
Comment (jedichien) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: done :)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java (hunk 19 above)
Comment (jedichien) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: I change it to `testPrintStreamWithProvidedKeyValueMapper` because I think use, test, in prefix to represent this is test case :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: I think renaming to `label` universally is okay. BTW for KTable its `print` function has been deprecated as its semantics is often confused by users, and we recommend people to use `KTable#toStream#print` instead. But it still worth just changing its javadoc.

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java (hunk 16 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: I think @dguy 's point is that, it is better to keep the `null` checks only in one function, and that function would better be as higher-level in the call trace (e.g. on the callers rather than callee). For this specific class, we can consolidate the `null` checks only in the `KStreamImpl`, and pass the default `mapper` / `label` / `writer` etc when the passed values are null, then we do not need to remember doing such further checkings in the lower level classes like `PrintForeachAction`.

So I'd say we can at least try enforcing such a pattern in this PR and gradually improve other functions other than `print` in future PRs.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: I think @dguy 's point is that in `ForeachPrintAction#apply` we do not need the `if-else` condition while in its `#close` we still maintain the `if-else` to special-handle `System.out` case?

**Review by guozhangwang — COMMENTED** [review]
Just trying to clarify some comments here, otherwise LGTM. @jedichien seems you have not pushed the latest commit addressing the previous comments yet though you mentioned some of them are `done`. Could you double check?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 9 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: ok, I will do it

File: streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java (hunk 16 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java]: ok, I will do it, thanks @guozhangwang :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 11 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: @guozhangwang Yes it is

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 13 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: ok :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 14 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: note that I just do it in local and not yet to PR, sorry for an ambiguous statement

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java (hunk 17 above)
Comment (jedichien) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: note that I just do it in local and not yet to PR, sorry for an ambiguous statement

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java (hunk 18 above)
Comment (jedichien) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: note that I just do it in local and not yet to PR, sorry for an ambiguous statement

**Comment by jedichien** [comment]
@guozhangwang I just do it in local and not yet to PR, sorry for an ambiguous statement

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 2 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:334]: Thanks @bbejeck for reminding me, I will add this in comment :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 3 above)
Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:356]: ok!

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6105/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6090/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6113/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6098/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 12 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: You only need to crate that instance once, right? It can be a member of the class

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java (hunk 19 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamPrintTest.java]: Hmm, i'm ok with using test, but we already know it is a test case. Also using `should` helps to describe what the class should do.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 20]
```diff
@@ -177,63 +177,146 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
 
     @Override
     public void print() {
-        print(null, null, null);
+        final KeyValueMapper<? super K, ? super V, String> mapper = new KeyValueMapper<K, V, String>() {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: I think we can extract this mapper to a field? It is is the same in all `print` and `writeAsText` methods, right?

**Review by dguy — COMMENTED** [review]
Thanks @jedichien, I've left a couple more comments

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6117/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
FAILURE
 3873 tests run, 0 skipped, 2 failed.
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6132/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by jedichien** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6118/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6133/
Test FAILed (JDK 7 and Scala 2.11).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 21]
```diff
@@ -760,8 +720,14 @@ public boolean test(final K1 key, final V1 value) {
                                         this.repartitionRequired);
     }
 
-
-
+    private <K, V> KeyValueMapper<K, V, String> createDefaultKeyValueMapper() {
+        return new KeyValueMapper<K, V, String>() {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: I think we should just create this as a field on the instance? I don't see any reason to create it everytime. Also makes the code cleaner

Comment (jedichien) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: ok!

**Comment by mjsax** [comment]
FYI: it the test fails again for the SSL test -- it's a know issue. Seems that the test is broken. No need to re-test again.

**Review by dguy — APPROVED** [review]
Thanks for the patch @jedichien, LGTM
I'm just going to run the tests again to make sure everything is all good.

**Comment by dguy** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6200/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6184/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
Thanks @jedichien, Merge to trunk!

**Comment by jedichien** [comment]
Thanks @dguy, @guozhangwang and @bbejeck for giving me lots of good advice :D 
This is my second PR, I cannot wait to make more contributions!!
BTW, I notice that this(and before) username in commit cannot link to my account, how should I fix it?

**Comment by guozhangwang** [comment]
@jedichien Thanks for your contributions!

Yes I'd recommend you fix your username (better be your real name in full) / email in your local github profile. Currently your github id profile is shown as `jameschien <jameschien@staff.ruten.com.tw>`.
